pkg/controllers/static/deprovisioning/controller.go

```diff
@@ -0,0 +1,179 @@
+	"sigs.k8s.io/karpenter/pkg/events"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	nodeutils "sigs.k8s.io/karpenter/pkg/utils/node"
+	nodepoolutils "sigs.k8s.io/karpenter/pkg/utils/nodepool"
+	"sigs.k8s.io/karpenter/pkg/utils/pretty"
+)
+
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	cluster       *state.Cluster
+	recorder      events.Recorder
+	clock         clock.Clock
+}
+
+func NewController(kubeClient client.Client, cluster *state.Cluster, recorder events.Recorder, cloudProvider cloudprovider.CloudProvider, clock clock.Clock) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		cluster:       cluster,
+		recorder:      recorder,
+		clock:         clock,
+	}
+}
+
+// Reconcile the resource
+func (c *Controller) Reconcile(ctx context.Context, np *v1.NodePool) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "deprovisioning.static")
+
+	if !nodepoolutils.IsManaged(np, c.cloudProvider) || np.Spec.Replicas == nil {
+		return reconcile.Result{}, nil
+	}
+
+	currentNodeClaims := helper.RunningNodesForNodePool(c.cluster, np)
+	desiredReplicas := lo.FromPtr(np.Spec.Replicas)
+	nodeClaimsToTerminate := currentNodeClaims - desiredReplicas
+
+	// Only handle scale down - scale up is handled by provisioning controller
+	if nodeClaimsToTerminate <= 0 {
+		return reconcile.Result{}, nil
+	}
+
+	log.FromContext(ctx).WithValues("NodePool", klog.KObj(np)).
+		Info("scaling down static nodepool", "current", currentNodeClaims, "desired", desiredReplicas, "toTerminate", nodeClaimsToTerminate)
+
+	// Get all active NodeClaims for this NodePool
+	// todo : cluster.Nodes() deepcopies all state nodes, we want to optimize to deep copy right nodes
+	allNodes := c.cluster.Nodes().Active()
+	npStateNodes := lo.Filter(allNodes, func(node *state.StateNode, _ int) bool {
+		return node.Labels()[v1.NodePoolLabelKey] == np.Name && node.NodeClaim != nil
+	})
+
+	// Get deprovisioning candidates
+	candidates := helper.GetDeprovisioningCandidates(ctx, c.kubeClient, np, npStateNodes, int(nodeClaimsToTerminate), c.clock)
+
+	scaleDownErrs := make([]error, len(candidates))
+	actualTerminatedCount := int64(0)
+	// Terminate selected NodeClaims
+	workqueue.ParallelizeUntil(ctx, len(candidates), len(candidates), func(i int) {
+		candidate := candidates[i]
+
+		if err := retry.OnError(retry.DefaultBackoff, func(err error) bool { return client.IgnoreNotFound(err) != nil }, func() error {
+			return c.kubeClient.Delete(ctx, candidate.NodeClaim)
+		}); err != nil && client.IgnoreNotFound(err) != nil {
+			log.FromContext(ctx).Error(err, "failed to delete NodeClaim", "NodeClaim", klog.KObj(candidate.NodeClaim))
+			scaleDownErrs[i] = err
+			return
+		}
+
+		atomic.AddInt64(&actualTerminatedCount, 1)
+
+		c.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, helper.TerminationReason)...)
```

Comment:
I'd consider a separate event since this isn't part of the graceful disruption mechanism. Currently this event is only emitted for graceful disruptions. I'm not just differentiating based on where it's implemented, but based on it's semantic differences:
- Doesn't respect disruption budgets
- Doesn't block on PDBs and `do-not-disrupt` pods

We don't emit this event for expiration either. If we want to add this event here, we should add it for all disruptions, or we should use a separate one. I'm happy with either, as long as we're consistent.

Reply:
Since deprovisioning is a user driven scale-down with explicit intent, I will remove this event.